### PR TITLE
perf(data-plane): avoid local batch copies

### PR DESCRIPTION
**Pull request by rohitrango**
# What does this PR do ?

Avoids unnecessary copies when fetching full or subset batches from the local data plane. It also preserves PackedTensor physical-segment sharing across replica broadcasts when that sharing reaches the worker intact, including the local data-plane path. The TQ storage codec remains unchanged.

Packed payloads are coalesced and moved to the collective device one field at a time, which bounds GPU staging memory for large multimodal batches.

Adds coverage that verifies full local fetches share tensor storage and replica broadcasts preserve PackedTensor row mappings, sharing metadata, preprocessing configuration, and physical tensors. A two-rank NCCL test covers CPU-to-GPU staging, CPU restoration, and int16 transport.

This PR is stacked on https://github.com/NVIDIA-NeMo/RL/pull/4106 via `rohit/packedtensor_padding`.

# Issues

None.

# Usage

No user-facing API changes.

# Before your PR is "Ready for review"

**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [x] Did you add or update any necessary documentation? The replica-broadcast scope is documented above.

# Additional Information

Validated on Slurm job `18295320` (`pool0-01445`, 8xH100):

```text
uv run --frozen --group test pytest --confcutdir=tests/unit/data_plane -q tests/unit/data_plane/test_leader_broadcast.py tests/unit/data_plane/test_local_sft.py
uv run --frozen --group test pytest --confcutdir=tests/unit/data -m "not mcore" -q tests/unit/data/test_multimodal_dict.py
```

Results: 17 data-plane tests passed, including the two-rank NCCL test; 59 PackedTensor tests passed. Ruff and Pyrefly passed for the changed implementation files.

## Previous PR and self-review

Previous PR: https://github.com/NVIDIA-NeMo/RL/pull/4091. A self-review was performed on the previous PR, and the resulting feedback was addressed in this PR.

### CODE REVIEW

File: nemo_rl/data_plane/adapters/local.py [hunk 1]
```diff
@@ -94,26 +94,28 @@ def _value_batch_size(value: Any) -> int:
...
-    return deepcopy(value)
+def _is_identity_indices(indices: list[int], batch_size: int) -> bool:
+    return len(indices) == batch_size and all(
+        index == position for position, index in enumerate(indices)
+    )
 
 
 def _select_batch_value(value: Any, indices: list[int]) -> Any:
+    if _is_identity_indices(indices, _value_batch_size(value)):
+        return value
```
Comment (terrykong) [inline nemo_rl/data_plane/adapters/local.py]: This fast path returns the object the store is holding, not a copy. Together with the other change in this PR, that lets a consumer move the store's own tensors onto the GPU.

The chain:

- <a href="https://github.com/NVIDIA-NeMo/RL/blob/2b9e7fc7777241454c6dbe2699c55825eeaf8caf/nemo_rl/data_plane/adapters/local.py#L390">`put_samples` now keeps the caller's object as-is</a> — `batch[name] = value`. On the base it was `batch[name] = _copy_batch_value(value)` (<a href="https://github.com/NVIDIA-NeMo/RL/blob/e347801924d50e0108660e39b5e8f9d35f2f283e/nemo_rl/data_plane/adapters/local.py#L388">base `local.py:388`</a>).
- This fast path hands that same object back to the fetcher.
- <a href="https://github.com/NVIDIA-NeMo/RL/blob/2b9e7fc7777241454c6dbe2699c55825eeaf8caf/nemo_rl/distributed/batched_data_dict.py#L1115-L1122">`BatchedDataDict.to()`</a> treats the two cases differently. For a plain tensor it rebinds (`self[k] = v.to(device)`), so the store is untouched. For a `PackedTensor` it calls <a href="https://github.com/NVIDIA-NeMo/RL/blob/2b9e7fc7777241454c6dbe2699c55825eeaf8caf/nemo_rl/data/multimodal_utils.py#L734-L736">`PackedTensor.to()`</a>, which does `self.tensors = [item.to(device) ...]; return self` — same object, new list, mutated in place. **That asymmetry is the trap**: plain tensors are safe here, `PackedTensor` is not.
- On the shipped path this is reachable. Energon VLM SFT is the only non-test writer (<a href="https://github.com/NVIDIA-NeMo/RL/blob/2b9e7fc7777241454c6dbe2699c55825eeaf8caf/nemo_rl/data/energon/sft_worker.py#L152">`sft_worker.py:152`</a>), it registers one partition per DP rank with `num_samples=batch_size`, so every fetch is the whole partition and always takes this path. Then `train` runs <a href="https://github.com/NVIDIA-NeMo/RL/blob/2b9e7fc7777241454c6dbe2699c55825eeaf8caf/nemo_rl/models/policy/dtensor_policy_worker.py#L662">`data.to("cuda")`</a>.

Net effect: the process-local store's segments get moved onto the GPU by its own consumer.

The clearest evidence that this matters is already in the file. <a href="https://github.com/NVIDIA-NeMo/RL/blob/2b9e7fc7777241454c6dbe2699c55825eeaf8caf/nemo_rl/data/multimodal_utils.py#L716-L733">`PackedTensor.to`'s own docstring</a> argues the bad case can't happen, and one of its three stated reasons is that *"worker-side device moves happen post-serialization on independent copies"*. That independence is exactly what the deleted copy provided. So this isn't "there's no written contract" — there is one, and this PR makes one of its premises false.

To keep this honest about scope:

- I could not build a case that produces wrong numbers. The problem is ownership and lifetime, not math.
- It doesn't pile up. The partition is fetched once and cleared at the end of the step (<a href="https://github.com/NVIDIA-NeMo/RL/blob/2b9e7fc7777241454c6dbe2699c55825eeaf8caf/examples/sft_v2.py#L265">`sft_v2.py:265`</a> → `clear_samples`).
- <a href="https://github.com/NVIDIA-NeMo/RL/blob/2b9e7fc7777241454c6dbe2699c55825eeaf8caf/nemo_rl/data_plane/adapters/local.py#L477">`save_checkpoint` pickles `partition.batch`</a>, so a checkpoint taken while a partition is live would write CUDA tensors.
- Today the only thing holding it back is <a href="https://github.com/NVIDIA-NeMo/RL/blob/2b9e7fc7777241454c6dbe2699c55825eeaf8caf/examples/sft_v2.py#L400">`sft_v2.py:400`</a>, which limits the local plane to Megatron, and the Megatron path slices first so `.to()` lands on a microbatch copy. The DTensor workers already make the unsafe top-level call. So this is one gate removal away from mattering.

The fix is one line and costs nothing:

```suggestion
    if not isinstance(value, PackedTensor) and _is_identity_indices(
        indices, _value_batch_size(value)
    ):
        return value
```

Falling through sends `PackedTensor` to `value.slice(...)`, which already shares the physical segments instead of copying the data (<a href="https://github.com/NVIDIA-NeMo/RL/blob/2b9e7fc7777241454c6dbe2699c55825eeaf8caf/nemo_rl/data/multimodal_utils.py#L792">`multimodal_utils.py:792`</a> and <a href="https://github.com/NVIDIA-NeMo/RL/blob/2b9e7fc7777241454c6dbe2699c55825eeaf8caf/nemo_rl/data/multimodal_utils.py#L812">`:812`</a>), so you keep the speedup. Measured locally on CPU, relative cost only: on a non-dedup `PackedTensor` of 64 rows / 86.7 MB, the gated `slice` costs ~0.001 ms against ~3.2 ms for the `deepcopy` this PR removes — three orders of magnitude cheaper than the thing being deleted, so this is not a partial revert.

To be clear, I'm not asking you to put the copy back — this one line is the whole fix. Plain tensors are already safe on this path: <a href="https://github.com/NVIDIA-NeMo/RL/blob/2b9e7fc7777241454c6dbe2699c55825eeaf8caf/nemo_rl/distributed/batched_data_dict.py#L1119">`batched_data_dict.py:1119`</a> rebinds the dict entry and `Tensor.to` returns a new tensor, so the store never sees the move. `PackedTensor` is the one type where the identity shortcut meets a `.to()` that mutates. Everything else the PR removes — which is the actual speedup — stays.

And to be clear about what else we're *not* saying: `rebuild_from_broadcast_parts` mutating `self` is fine, and so is `PackedTensor.to()` in general. In-place mutation is fine when the object belongs to the caller making the call; it's a problem only when the object is shared with someone else, which is what the deleted copy used to prevent.

One small thing in the same area, since it's the same idea: the <a href="https://github.com/NVIDIA-NeMo/RL/blob/2b9e7fc7777241454c6dbe2699c55825eeaf8caf/nemo_rl/data/multimodal_utils.py#L482-L485">`broadcast_parts` docstring</a> says "new semantic state does not require a parallel broadcast schema update". That's too broad — four lines down the code raises if the state holds a tensor. Please say "new **non-tensor** state". One word.

File: tests/unit/data_plane/test_leader_broadcast.py [hunk 2]
```diff
@@ -138,6 +141,50 @@ def _round_trip_body(rank: int):
     assert torch.equal(packed.as_tensor(), expected.as_tensor())
 
 
+def _deduplicated_round_trip_body(rank: int):
```
Comment (terrykong) [inline tests/unit/data_plane/test_leader_broadcast.py]: This helper compares `.tensors` positionally, which is right. The comment a few lines above, at <a href="https://github.com/NVIDIA-NeMo/RL/blob/2b9e7fc7777241454c6dbe2699c55825eeaf8caf/tests/unit/data_plane/test_leader_broadcast.py#L133-L134">L133-134</a>, now says something that is no longer true:

> Compare on logical rows, not `.tensors`: `from_wire` returns segments flat with a CSR row map, so an empty row contributes no entry there.

This PR takes `from_wire` out of that path — `grep -c "from_wire\|to_wire" nemo_rl/data_plane/worker_mixin.py` returns 0 at this head. The follower now rebuilds through `rebuild_from_broadcast_parts`, which keeps `None` rows where they were, so the follower's `.tensors` matches the leader's element for element. This PR's own new test depends on that: it asserts <a href="https://github.com/NVIDIA-NeMo/RL/blob/2b9e7fc7777241454c6dbe2699c55825eeaf8caf/tests/unit/data_plane/test_leader_broadcast.py#L181">`packed.tensors[1] is None`</a>.

Either drop the comment, or compare `.tensors` positionally there the way this helper does.

(Anchored here rather than on L133-134, since those lines are not in the diff and so cannot carry a comment.)

File: tests/unit/data_plane/test_leader_broadcast.py [hunk 3]
```diff
@@ -192,6 +281,34 @@ def test_leader_broadcast_keeps_media_free_packed_key(tmp_path):
...
+
+def test_packed_tensor_broadcast_rejects_tensor_header_state():
+    packed = _packed(_pixel_rows())
+    packed.__dict__["cached_tensor"] = torch.ones(1)
+
+    with pytest.raises(TypeError, match="header must be tensor-free"):
+        packed.broadcast_parts()
+
+
+def test_packed_tensor_broadcast_rejects_oversized_payload():
```
Comment (terrykong) [inline tests/unit/data_plane/test_leader_broadcast.py]: `broadcast_parts` adds three raises, and this PR covers two of them: the tensor-bearing header (<a href="https://github.com/NVIDIA-NeMo/RL/blob/2b9e7fc7777241454c6dbe2699c55825eeaf8caf/tests/unit/data_plane/test_leader_broadcast.py#L296">`:296`</a>) and the oversized payload (this test). The mixed-dtype guard (<a href="https://github.com/NVIDIA-NeMo/RL/blob/2b9e7fc7777241454c6dbe2699c55825eeaf8caf/nemo_rl/data/multimodal_utils.py#L491">`multimodal_utils.py:491`</a>) and the mixed-device guard (<a href="https://github.com/NVIDIA-NeMo/RL/blob/2b9e7fc7777241454c6dbe2699c55825eeaf8caf/nemo_rl/data/multimodal_utils.py#L494">`:494`</a>) have no test — `grep -rn "one dtype for broadcast" tests/` finds nothing.

The dtype one is worth covering: the payload is a single flat `torch.cat`, so a mixed-dtype column would be read back under one wire dtype on the follower. Here's the test — I ran it against this PR unmodified and it passes:

```python
def test_packed_tensor_broadcast_rejects_mixed_dtype_segments():
    packed = PackedTensor(
        [torch.ones(2, dtype=torch.float32), torch.ones(2, dtype=torch.int16)],
        dim_to_pack=0,
    )

    with pytest.raises(TypeError, match="one dtype for broadcast"):
        packed.broadcast_parts()
```

Low priority — the guard already works, so this pins it rather than fixing anything. We're not asking for the mixed-device test; that needs two devices, and one GPU-gated test in this file is enough.

File: tests/unit/data_plane/test_local_sft.py [hunk 4]
```diff
@@ -87,6 +87,39 @@ def test_local_round_trip_preserves_tensor_and_packed_tensor_fields() -> None:
...
+def test_local_full_fetch_shares_tensor_storage() -> None:
+    client = _client()
+    _register(client)
+    input_ids = torch.tensor([[1, 2, 0], [3, 4, 5]])
+    pixel_rows = [torch.full((1, 2), 1.0), torch.full((2, 2), 2.0)]
+    pixels = PackedTensor(
+        pixel_rows,
+        dim_to_pack=0,
+        preprocess_mode="pad_to_max_shape",
+    ).enable_deduplication()
```
Comment (terrykong) [inline tests/unit/data_plane/test_local_sft.py:99]: Two of this test's three assertions pass at the base commit as well as at head, so they don't pin what the PR changed. The cause is `.enable_deduplication()` on this line: that is the one path where <a href="https://github.com/NVIDIA-NeMo/RL/blob/2b9e7fc7777241454c6dbe2699c55825eeaf8caf/nemo_rl/data/multimodal_utils.py#L600-L631">`PackedTensor.__deepcopy__`</a> already shared the physical tensors, so the segments were shared before this PR too.

Measured through the real `put_samples` → `get_data` → `materialize_local` path at both commits:

```
                                                     BASE    HEAD
dedup segment-identity (what this test asserts)      True    True
NON-dedup segment-identity                          False    True
input_ids data_ptr shared                           False    True
```

Only the `data_ptr` assertion on line 117 tells the two apart. Dropping deduplication makes the segment assertions flip False→True across the change, which is what they were clearly meant to do:

```suggestion
    )
```

One caution if you want to check this yourself: don't judge it by running the test at the base commit. It fails there at line 117 first, so pytest never reaches the segment assertions — a red run is equally consistent with those assertions being meaningful or vacuous. That is a property of the assertion ordering, so it stays true after this fix. The per-assertion table above is the only evidence that settles it.

Applied together with the `local.py` suggestion, both changed test files still give 16 passed / 1 skipped (the skip is the NCCL test, which needs two CUDA devices). Applied alone, `test_local_sft.py` gives 9 passed.

**Review by terrykong — COMMENTED** [review]
The main win here is real: deduplicated image segments now survive the replica broadcast. The old path went through `to_wire`, which emits one row per *logical* row, so an image shared by several samples crossed the wire once per sample and every receiving rank rebuilt that many copies. [`broadcast_parts`](https://github.com/NVIDIA-NeMo/RL/blob/4fb9fbca541bcd62668651039f907acae4871050/nemo_rl/data/multimodal_utils.py#L473-L545) sends the object with its tensors removed, plus each distinct segment once, so the sharing the leader already did is still there on the other side.

Dropping the `empty_packed` descriptor kind is a second, quieter win worth calling out: the old branch rebuilt followers by hand as `PackedTensor([None] * n_rows, ...)` and silently lost `_row_offsets` / `_segment_indices` / `_segment_provenance` ([base `worker_mixin.py:241-251`](https://github.com/NVIDIA-NeMo/RL/blob/e347801924d50e0108660e39b5e8f9d35f2f283e/nemo_rl/data_plane/worker_mixin.py#L241-L251), linked at the base since it is gone at head). That was a real loss of state, not just extra branches.

**CI.** No lint or unit-test job has run on this head. The always-on gates (DCO, copyright, secrets, semantic PR title, submodule, release) passed, but `gh pr checks 4107` shows no `Lint check` and no L0 unit job. Two things unblock it: the PR has no labels and needs a `CI:*` one, and someone has to post `/ok to test 4fb9fbca541bcd62668651039f907acae4871050` (full SHA, re-posted after every push). Your description says you ran the data-plane and multimodal tests on 8xH100, so this is "please trigger CI", not "this is untested".

**What we checked locally.** Both changed test files: 16 passed, 1 skipped. The skip is the NCCL case, which needs two CUDA devices — so this is "the CPU suite passes here", not "the tests pass". Ruff is clean via `uvx ruff@0.9.9`, which is not the same as CI's `Lint check` (that also runs pyrefly and a whitelist rule).

**Blast radius, since the three files differ.** [`worker_mixin.py`](https://github.com/NVIDIA-NeMo/RL/blob/4fb9fbca541bcd62668651039f907acae4871050/nemo_rl/data_plane/worker_mixin.py#L69) is shared with the DTensor v2 worker through `TQWorkerMixin`, so the broadcast rewrite reaches beyond the Energon path — that is the part worth the most scrutiny. [`local.py`](https://github.com/NVIDIA-NeMo/RL/blob/4fb9fbca541bcd62668651039f907acae4871050/nemo_rl/data_plane/adapters/local.py#L104-L105) does not: the local store is reachable only from SFTv2, which [raises unless Megatron is enabled](https://github.com/NVIDIA-NeMo/RL/blob/4fb9fbca541bcd62668651039f907acae4871050/nemo_rl/algorithms/sft_v2.py#L399-L400). And `multimodal_utils.py` is purely additive — one widened import plus two new methods, no existing method touched — so current `PackedTensor` callers are unaffected.

One comment below, on a test this PR adds. Nothing here blocks the merge.

**Review by terrykong — DISMISSED** [review]

**Review by terrykong — APPROVED** [review]
